parser/types.go

```diff
@@ -119,29 +119,46 @@ func incompletePipeExpr(call interface{}, text []byte, pos position) (*ast.PipeE
 func memberexprs(head, tail interface{}, text []byte, pos position) (ast.Expression, error) {
 	res := head.(ast.Expression)
 	for _, prop := range toIfaceSlice(tail) {
-		res = &ast.MemberExpression{
-			Object:   res,
-			Property: prop.(ast.Expression),
-			BaseNode: base(text, pos),
+		switch expr := prop.(type) {
+		case *memberExpression:
+			res = &ast.MemberExpression{
+				Object:   res,
+				Property: expr.property,
+				BaseNode: base(text, pos),
+			}
+		case *indexExpression:
+			if lit, ok := expr.key.(*ast.StringLiteral); ok {
+				res = &ast.MemberExpression{
+					Object:   res,
+					Property: &ast.Identifier{Name: lit.Value},
```

Comment:
I would like to preserve the invariant that an Identifier's Name is always a valid identifier. Using the string literal value as the Name breaks that. See my comment above about using *Identifier vs Expression as the Property.  

I think a change to the parser grammar is required to make this work.

There should be three production rules:

* One for `Expr "." Identifier` -> MemberExpression
* One for `Expr "[" StringLiteral "]"` -> MemberExpression
* One for `Expr "[" Expression "]" ` -> IndexExpression (the integer literal case fall into this one)

The StringLiteral is more specific than an Expression so it should appear first. This should remove the need to have the special `memberExpression` and `indexExpression` types in the parser.